dev-tools/xtask/src/a4x2_deploy.rs

```diff
@@ -0,0 +1,472 @@
+            // (and perhaps others?), to help diagnose a4x2 launch failures, and
+            // provide context alongside live tests.
+
+            // If we are in CI, we do not need to tear down, because the system
+            // state is about to be wiped from existence anyway.
+            if !env.in_ci {
+                // Outside of CI, we do not tear down if we are just running the
+                // start command, or if the user requested to leave a4x2 up
+                // after running the tests.
+                match args.command {
+                    DeployCommand::RunTests(RunTestsArgs {
+                        leave_running: false,
+                        ..
+                    }) => {
+                        // We ignore any error here, because the result we
+                        // actually want to produce is whether tests passed
+                        let _ = teardown_a4x2(&sh, &env);
+                    }
+
+                    _ => {
+                        // If we are leaving a4x2 up, we ought to be nice and
+                        // print some information to the user so they can get
+                        // into the system
+                        print_a4x2_access_info(&sh, &env);
+                    }
+                }
+            }
+
+            // Unwrap the launch/tests error now, if there was one
+            result?;
+        }
+    }
+
+    Ok(())
+}
+
+fn unpack_a4x2(sh: &Shell, env: &Environment) -> Result<()> {
+    cmd!(sh, "banner 'unpack'").run()?;
+    let tgz_path = &env.a4x2_package_tar;
+
+    if !tgz_path.try_exists()? {
+        bail!("a4x2-package bundle does not exist at {}, did you run `cargo xtask a4x2-package`?", tgz_path);
+    }
+
+    cmd!(sh, "tar -xvzf {tgz_path}").run()?;
+
+    if !env.a4x2_dir.try_exists()? {
+        bail!("extracting a4x2-package bundle did not result in a4x2-package-out/ existing");
+    }
+
+    let a4x2_dir = &env.a4x2_dir;
+    cmd!(sh, "tar -cf sled-common.tar -C {a4x2_dir}/cargo-bay/sled-common ./")
+        .run()?;
+    cmd!(sh, "tar -xf sled-common.tar -C {a4x2_dir}/cargo-bay/g0").run()?;
+    cmd!(sh, "tar -xf sled-common.tar -C {a4x2_dir}/cargo-bay/g1").run()?;
+    cmd!(sh, "tar -xf sled-common.tar -C {a4x2_dir}/cargo-bay/g2").run()?;
+    cmd!(sh, "tar -xf sled-common.tar -C {a4x2_dir}/cargo-bay/g3").run()?;
+
+    Ok(())
+}
+
+fn prepare_to_launch_a4x2(sh: &Shell, env: &Environment) -> Result<()> {
+    cmd!(sh, "banner 'prepare'").run()?;
+
+    // TODO could move these into the a4x2-package bundle.
+    // mostly of matter of having a consistent version of propolis.
+    exec_remote_script(&sh, "https://raw.githubusercontent.com/oxidecomputer/falcon/main/get-ovmf.sh")?;
+    exec_remote_script(&sh, "https://raw.githubusercontent.com/oxidecomputer/falcon/main/get-propolis.sh")?;
+
+    // Generate an ssh key we will use to log into the sleds.
+    cmd!(sh, "ssh-keygen -t ed25519 -N '' -f a4x2-ssh-key").run()?;
```

Comment:
Ooh, (I think) I like using a one-off key here.  But it also needs to be easy for folks to ssh in after the fact.  Should we give folks a command to load it into their agent?

Reply:
Or provide an ssh command to copy-paste.